### DISCUSSION TITLE
Health timeout Issue

### DISCUSSION THREAD

**Discussion by Chrisdup8710** [ptarmiganlabs/butler-sos#687] [Q&A]
Good day

I've recently implemented Butler-SOS on our environment where we have 2 central node, 2 presentation nodes (Proxy running through main central), and 7 reload nodes.

I'm currently struggling with the health check call to 6 of the servers where Butler is reporting that the requests keep timing out. Though sometimes those requests do go through, very intermittently though. Any ideas of where I should start looking for the issue?

I must also add that we do get some of the log events coming through from these servers.

**Comment by mountaindude** [comment]
Difficult to say what's going on without knowing more about your environment.

It's a fairly sizeable Sense setup, but nothing Butler and Butler SOS cannot handle. 



Some thoughts/questios:

- Where is Butler SOS running? Close to the Sense servers (from a network perspective)? Is there high load on the server where Butler SOS is running?
- Any errors or warnings in the Butler SOS log files?
- More Sense nodes mean that Butler SOS has more to do (obviously).
- The polling interval may be too short. This could in theory lead to Butler SOS not being able to capture all data until it's time to gather the next set of data, eventually leading to things timing out and loss of metrics. 
   - I would expect errors in the Butler SOS logs in this case
   - Polling for new data every 30 seconds should be fine, I'd say. Has worked well in other similar sized Sense environments. But you could try increasing poll time to a minute just to test the hypothesis.
   - Poll time is controlled via the config file setting `Butler-SOS.serversToMonitor.pollingInterval`.  Remember to restart Butler SOS after making changes to the config file!
- Network flakiness could be a reason to the symptoms you see. Unlikely maybe, but I did once see similar things in a network where there turned out to be routing issues. Some packets took the looong way to the destination (many timed out en route) and some packets were routed as planned.

**Comment by Chrisdup8710** [comment]
Hi Goran

Thanks for the quick reply!

My comments on your questions below:

On Wed, Dec 13, 2023 at 9:00 AM Göran Sander ***@***.***>
wrote:

> Difficult to say what's going on without knowing more about your
> environment.
>
> It's a fairly sizeable Sense setup, but nothing Butler and Butler SOS
> cannot handle.
>
> Some thoughts/questios:
>
>    - Where is Butler SOS running? Close to the Sense servers (from a
>    network perspective)? Is there high load on the server where Butler SOS is
>    running? I'm running Butler-SOS on my failover central node. It is a
>    passive node so not thinking high load would be the issue here.
>    - Any errors or warnings in the Butler SOS log files? The only errors
>    coming through is as below:
>
>  2023-12-13T03:36:57.937Z error: HEALTH: Error when calling health check
API: AxiosError: timeout of 5000ms exceeded
It was also difficult to figure out which servers were struggling, so I
added the ${host} in the error code to give the error as below:
2023-12-13T07:30:54.387Z error: HEALTH: Error when calling health check API
for pcdwqs9zatcwi.vodacom.corp:4747: AxiosError: timeout of 10000ms exceeded
As you can see I played with the timeout value, went up to 15000, but it
doesn't seem like that had an effect.

>
>    - More Sense nodes mean that Butler SOS has more to do (obviously). I
>    must say it doesn't look like we are struggling here
>    - The polling interval may be too short. This could in theory lead to
>    Butler SOS not being able to capture all data until it's time to gather the
>    next set of data, eventually leading to things timing out and loss of
>    metrics.
>       - I would expect errors in the Butler SOS logs in this case
>       - Polling for new data every 30 seconds should be fine, I'd say.
>       Has worked well in other similar sized Sense environments. But you could
>       try increasing poll time to a minute just to test the hypothesis.
>       - Poll time is controlled via the config file setting
>       Butler-SOS.serversToMonitor.pollingInterval. Remember to restart
>       Butler SOS after making changes to the config file!
>    - Network flakiness could be a reason to the symptoms you see.
>    Unlikely maybe, but I did once see similar things in a network where there
>    turned out to be routing issues. Some packets took the looong way to the
>    destination (many timed out en route) and some packets were routed as
>    planned. Are you aware of a way to check if there are issues like this
>    in my case?
>
>
Thanks again for the reply and looking forward to chatting some more. I am
looking to deploy Butler (To integrate to Pagerduty) and Butler-Spyglass
(for lineage, we have a store of about 3Tb)

>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ptarmiganlabs/butler-sos/discussions/687#discussioncomment-7838734>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHL6DU22WQYMOUFSKFLDQTTYJFHBTAVCNFSM6AAAAABASSLOACVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM3TQMZYG4ZTI>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***
> com>
>


-- 
Chris du Plessis
0825230782
***@***.***

**Reply by mountaindude** [reply]
Hmm, if Axios times out that indicates it didn't get a response back from the remote QS server in time. 
That speaks against the theory of Butler SOS not being able to keep up with a (too) high polling frequency.

But if that Axios error only happens _sometimes_ for a specific remote QS host... then I don't know what's going on.
Really hard to debug without access to the actual environment.

Re integration between Butler and PagerDuty that's been on my radar for some time too.
The New Relic integration was higher priority though, but the general concepts developed there can be applied also to PagerDuty. 
Except that PD won't have a need for time-series metrics in the same way that NR does.

Re Butler Spyglass, that's an old goldie... :)
It does the job when it comes to extracting lineage data, the challenge is then to stick that data together into something that makes sense for your use case. 
You would think that "data lineage" means the same to different organisations/teams... not so much. 😄 
So Butler Spyglass' approach is to provide the raw lineage info, but not a ready-to-use UI on top of it.
That said, I've done some work recently around providing full source-via-QVDs-to-charts-in-apps lineage. 
Possible to do, but not without challenges. 

Btw, I am rolling off my current gig in a week, available for work starting Jan/Feb next year. 
In case you guys need help with these things... Ping me at info <at> ptarmiganlabs <dot>  com if so.

**Reply by mountaindude** [reply]
@Chrisdup8710 

I thought a bit more about different ways of getting Qlik Sense alerts over to PagerDuty.

I can see two ways of making this happen (using Butler and Butler SOS):

1. Add native PagerDuty support in Butler and Butler SOS. Would take some development work, but would also result in the cleanest and most flexible solution.
2. Use the fact that Butler and Butler SOS already today can push data into InfluxDb, which in turn can be accessed very nicely from Grafana. Grafana in turn has a PagerDuty plugin in its standard alerting feature... 

Given the flexibility of Grafana's alerting feature that could actually be a pretty nice solution!

**Reply by Chrisdup8710** [reply]
@mountaindude 

Option 1 would be amazing! and honestly would love to be involved somehow!

Option 2 was what I was aiming for as I understood that to be the route to Pagerduty, and what I'm attempting to deploy now.

Thanks for the offer to help earlier. I'll send you a mail

**Comment by Chrisdup8710** [comment]
I was testing SOS on one server and wanted to deploy it on our central node as I know the port are open between that central node and the other nodes (trying to fix the earlier issue where engine health data was not coming through). 

Now the SOS instance there won't start with error:
c:\tools\butler-sos\src>node butler-sos.js
internal/modules/cjs/loader.js:638
    throw err;
    ^

Error: Cannot find module 'node:http'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:636:15)
    at Function.Module._load (internal/modules/cjs/loader.js:562:25)
    at Module.require (internal/modules/cjs/loader.js:692:17)
    at require (internal/modules/cjs/helpers.js:25:18)
    at Object.<anonymous> (c:\tools\butler-sos\node_modules\fastify\fastify.js:6:14)
    at Module._compile (internal/modules/cjs/loader.js:778:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:789:10)
    at Module.load (internal/modules/cjs/loader.js:653:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:593:12)
    at Function.Module._load (internal/modules/cjs/loader.js:585:3)

Any ideas? Looks like it is trying to find a module node:http, which I can't install running npm for some reason...

c:\tools\butler-sos\src>npm i node:http
npm ERR! code EUNSUPPORTEDPROTOCOL
npm ERR! Unsupported URL Type "node:": node:http

npm ERR! A complete log of this run can be found in: C:\Users\qliksens\AppData\Local\npm-cache\_logs\2023-12-14T04_35_01_457Z-debug-0.log

From what I've seen looking around, I should try http-node, and http though no success trying to install that either...

though just running npm i goes through fine:
c:\tools\butler-sos\src>npm i

up to date in 2m

120 packages are looking for funding
  run `npm fund` for details

**Reply by mountaindude** [reply]
Oh, you are using the actual node.js code. 

That should work too of course, but using the pre-compiled binary might be easier. You're then using a known-to-be-working version, eliminating possible npm issues etc.

Just for the record, it should be enough doing a `npm install` in the folder where the package.json file resides. 
That should download and install all needed dependencies.
But again - using the pre-compiled binary gets around that.

**Reply by Chrisdup8710** [reply]
Thanks @mountaindude! I've got it running now.

I was wondering about something, can I use the appenders you set up for butler in the scheduler appender file for butler-sos? Or should I run butler separately?

**Reply by mountaindude** [reply]
Great!

Yes you can (should!) use the same xml files for Butler and Butler SOS.

I'll see if I can dig up a sample file for you.
It's pretty straightforward though, just merge the files for Butler and Butler SOS into a single xml file.

**Reply by Chrisdup8710** [reply]
Great ok I'll see what I get if I merge those.

Also wanted to find out, the service monitoring you have in butler, can that be done in SOS?

**Comment by mountaindude** [comment]
Nope, service monitoring is in Butler only.

Major rework on that one in Butler version that was released last night,
btw

**Reply by Chrisdup8710** [reply]
Ah ok perfect. Will set it up quickly

**Comment by mountaindude** [comment]
Here is a sample XML log appender file for the Sense scheduler.

It has sections for both Butler and Butler SOS., with the assumption that both tools run on a server with IP `10.11.12.13`.
Note the different UDP ports for Butler and Butler SOS, `9998` and `9996`, respectively.

The file should be usable as-is (change the IP and possibly UDP ports). 
Some ppl tweak the filter strings to match on other/additional strings in the Sense logs, but that can/should only be explored after you have things working, IMHO.


```xml
<?xml version="1.0" encoding="UTF-8"?>
<configuration>
    <!-- Appender for detecting reload task failures. Only the last of potentially several retries is reported -->
    <appender name="TaskFailureLogger" type="log4net.Appender.UdpAppender">
        <filter type="log4net.Filter.StringMatchFilter">
            <param name="stringToMatch" value="Max retries reached" />
        </filter>
        <filter type="log4net.Filter.DenyAllFilter" />
        <param name="remoteAddress" value="10.11.12.13" />
        <param name="remotePort" value="9998" />
        <param name="encoding" value="utf-8" />
        <layout type="log4net.Layout.PatternLayout">
            <converter>
                <param name="name" value="hostname" />
                <param name="type" value="Qlik.Sense.Logging.log4net.Layout.Pattern.HostNamePatternConverter" />
            </converter>
            <param name="conversionpattern" value="/scheduler-reload-failed/;%hostname;%property{TaskName};%property{AppName};%property{User};%property{TaskId};%property{AppId};%date;%level;%property{ExecutionId};%message" />
        </layout>
    </appender>

    <!-- Appender for detecting aborted reloads -->
    <appender name="AbortedReloadTaskLogger" type="log4net.Appender.UdpAppender">
        <filter type="log4net.Filter.StringMatchFilter">
            <param name="stringToMatch" value="Execution State Change to Aborting" />
        </filter>
        <filter type="log4net.Filter.DenyAllFilter" />
        <param name="remoteAddress" value="10.11.12.13" />
        <param name="remotePort" value="9998" />
        <param name="encoding" value="utf-8" />
        <layout type="log4net.Layout.PatternLayout">
            <converter>
                <param name="name" value="hostname" />
                <param name="type" value="Qlik.Sense.Logging.log4net.Layout.Pattern.HostNamePatternConverter" />
            </converter>
            <param name="conversionpattern" value="/scheduler-reload-aborted/;%hostname;%property{TaskName};%property{AppName};%property{User};%property{TaskId};%property{AppId};%date;%level;%property{ExecutionId};%message" />
        </layout>
    </appender>

    <!-- Appender for detecting successful reload tasks -->
    <appender name="ReloadTaskSuccessLogger" type="log4net.Appender.UdpAppender">
        <filter type="log4net.Filter.StringMatchFilter">
            <param name="stringToMatch" value="Reload complete" />
        </filter>
        <filter type="log4net.Filter.DenyAllFilter" />
        <param name="remoteAddress" value="10.11.12.13" />
        <param name="remotePort" value="9998" />
        <param name="encoding" value="utf-8" />
        <layout type="log4net.Layout.PatternLayout">
            <converter>
                <param name="name" value="hostname" />
                <param name="type" value="Qlik.Sense.Logging.log4net.Layout.Pattern.HostNamePatternConverter" />
            </converter>
            <param name="conversionpattern" value="/scheduler-reloadtask-success/;%hostname;%property{TaskName};%property{AppName};%property{User};%property{TaskId};%property{AppId};%date;%level;%property{ExecutionId};%message" />
        </layout>
    </appender>


    <!-- Generic appender for detecting warnings and errors -->
    <appender name="LogEvent" type="log4net.Appender.UdpAppender">
        <param name="threshold" value="warn" />
        <param name="remoteAddress" value="10.11.12.13" />
        <param name="remotePort" value="9996" />
        <param name="encoding" value="utf-8" />
        <layout type="log4net.Layout.PatternLayout">
            <converter>
                <param name="name" value="rownum" />
                <param name="type" value="Qlik.Sense.Logging.log4net.Layout.Pattern.CounterPatternConverter" /> 
            </converter> 
            <converter>
                <param name="name" value="hostname" />
                <param name="type" value="Qlik.Sense.Logging.log4net.Layout.Pattern.HostNamePatternConverter" />
            </converter>
            <converter>
                  <param name="name" value="longIso8601date" /> 
                  <param name="type" value="Qlik.Sense.Logging.log4net.Layout.Pattern.Iso8601TimeOffsetPatternConverter" /> 
            </converter>
            <converter> 
                  <param name="name" value="user" /> 
                  <param name="type" value="Qlik.Sense.Logging.log4net.Layout.Pattern.ServiceUserNameCachedPatternConverter" /> 
            </converter> 
            <converter> 
                  <param name="name" value="encodedmessage" /> 
                  <param name="type" value="Qlik.Sense.Logging.log4net.Layout.Pattern.EncodedMessagePatternConverter" /> 
            </converter> 
            <converter> 
                  <param name="name" value="encodedexception" /> 
                  <param name="type" value="Qlik.Sense.Logging.log4net.Layout.Pattern.EncodedExceptionPatternConverter" /> 
            </converter>
            <param name="conversionpattern" value="/qseow-scheduler/;%rownum{9999};%longIso8601date;%date;%level;%hostname;%logger;%user;%encodedmessage;%encodedexception;%property{UserDirectory};%property{UserId};%property{User};%property{TaskName};%property{AppName};%property{TaskId};%property{AppId};%property{ExecutionId}" />
        </layout>
    </appender>

    <!-- Send UDP message to Butler SOS on warnings and errors -->
    <logger name="Service">
        <appender-ref ref="LogEvent" />
    </logger>
    <logger name="System">
        <appender-ref ref="LogEvent" />
    </logger>

    <!-- Send message to Butler on task failure -->
    <!-- Send message to Butler on task abort -->
    <logger name="System.Scheduler.Scheduler.Master.Task.TaskSession">
        <appender-ref ref="TaskFailureLogger" />
        <appender-ref ref="AbortedReloadTaskLogger" />
    </logger>

    <!-- Send message to Butler on reload task success -->
    <logger name="System.Scheduler.Scheduler.Slave.Tasks.ReloadTask">
        <appender-ref ref="ReloadTaskSuccessLogger" />
    </logger>
    
</configuration>
```

**Comment by mountaindude** [comment]
By the way, you may find the blogs at https://ptarmiganlabs.com of interest too. 
Quite a few SenseOps related articles there. Feel free to sign up for the newsletter too...Thx! 🙏

**Comment by mountaindude** [comment]
Closing old ticket
